File: src/main/java/io/github/takenoha/towerdefense/paper/RaidSealTagger.java
package io.github.takenoha.towerdefense.paper;

import io.github.takenoha.towerdefense.domain.StageWaveSchedule;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.OptionalLong;
import java.util.UUID;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.persistence.PersistentDataContainer;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.plugin.Plugin;

/** Creates and validates the non-stackable, database-backed raid-start token. */
public final class RaidSealTagger {
    public static final int ITEM_VERSION = 1;
    public static final long FOUNDATION_STAGE = 1L;
    public static final Material ITEM_MATERIAL = Material.valueOf(
            RaidSealMaterialPolicy.CURRENT_MATERIAL);
    public static final Material LEGACY_ITEM_MATERIAL = Material.valueOf(
            RaidSealMaterialPolicy.LEGACY_MATERIAL);

    private final NamespacedKey markerKey;
    private final NamespacedKey versionKey;
    private final NamespacedKey sealIdKey;
    private final NamespacedKey stageLevelKey;

    public RaidSealTagger(Plugin plugin) {
        Objects.requireNonNull(plugin, "plugin");
        markerKey = new NamespacedKey(plugin, "raid_seal");
        versionKey = new NamespacedKey(plugin, "raid_seal_version");
        sealIdKey = new NamespacedKey(plugin, "raid_seal_id");
        stageLevelKey = new NamespacedKey(plugin, "raid_seal_stage");
    }

    /** Template used by the stage-1 recipe; a craft event replaces its empty UUID marker. */
    public ItemStack recipeTemplate() {
        return recipeTemplate(FOUNDATION_STAGE);
    }

    /** Template used by a stage-specific recipe before craft-time UUID registration. */
    public ItemStack recipeTemplate(long stageLevel) {
        return createTemplate(stageLevel);
    }

    public ItemStack create(UUID sealId, long stageLevel) {
        Objects.requireNonNull(sealId, "sealId");
        requireStage(stageLevel);
        ItemStack item = createTemplate(stageLevel);
        ItemMeta meta = requireMeta(item);
        PersistentDataContainer data = meta.getPersistentDataContainer();
        data.set(sealIdKey, PersistentDataType.STRING, sealId.toString());
        data.set(stageLevelKey, PersistentDataType.LONG, stageLevel);
        meta.displayName(Component.text("襲撃の印", NamedTextColor.GOLD));
        meta.lore(List.of(
                Component.text("ステージ" + stageLevel + "の防衛戦を開始します", NamedTextColor.GRAY),
                Component.text("右クリックでチームのコアへ使用", NamedTextColor.GRAY),
                Component.text("真正性はサーバーの永続データで検証されます", NamedTextColor.DARK_GRAY)));
        item.setItemMeta(meta);
        item.setAmount(1);
        return item;
    }

    public Optional<RaidSealItemIdentity> read(ItemStack item) {
        if (item == null || !isSupportedMaterial(item.getType()) || item.getAmount() != 1) {
            return Optional.empty();
        }
        ItemMeta meta = item.getItemMeta();
        return meta == null ? Optional.empty() : read(meta.getPersistentDataContainer());
    }

    /** Returns whether a valid seal still uses the pre-UX ENDER_EYE material. */
    public boolean isLegacyMaterial(ItemStack item) {
        return item != null
                && item.getType() == LEGACY_ITEM_MATERIAL
                && read(item).isPresent();
    }

    public boolean isRecipeTemplate(ItemStack item) {
        return templateStage(item).isPresent();
    }

    /** Reads the stage encoded in a registered recipe result. */
    public OptionalLong templateStage(ItemStack item) {
        if (item == null || !isSupportedMaterial(item.getType()) || item.getAmount() != 1) {
            return OptionalLong.empty();
        }
        ItemMeta meta = item.getItemMeta();
        if (meta == null) {
            return OptionalLong.empty();
        }
        PersistentDataContainer data = meta.getPersistentDataContainer();
        Integer version = data.get(versionKey, PersistentDataType.INTEGER);
        Byte marker = data.get(markerKey, PersistentDataType.BYTE);
        if (marker == null
                || marker != 1
                || version == null
                || version != ITEM_VERSION
                || data.get(sealIdKey, PersistentDataType.STRING) != null) {
            return OptionalLong.empty();
        }
        Long stage = data.get(stageLevelKey, PersistentDataType.LONG);
        if (stage == null) {
            return OptionalLong.empty();
        }
        try {
            StageWaveSchedule.requireValidStageLevel(stage);
            return OptionalLong.of(stage);
        } catch (IllegalArgumentException invalidStage) {
            return OptionalLong.empty();
        }
    }

    public boolean hasSealId(ItemStack item, UUID sealId) {
        Objects.requireNonNull(sealId, "sealId");
        return read(item).map(identity -> identity.sealId().equals(sealId)).orElse(false);
    }

    private ItemStack createTemplate(long stageLevel) {
        requireStage(stageLevel);
        ItemStack item = new ItemStack(ITEM_MATERIAL, 1);
        ItemMeta meta = requireMeta(item);
        PersistentDataContainer data = meta.getPersistentDataContainer();
        data.set(markerKey, PersistentDataType.BYTE, (byte) 1);
        data.set(versionKey, PersistentDataType.INTEGER, ITEM_VERSION);
        data.set(stageLevelKey, PersistentDataType.LONG, stageLevel);
        meta.displayName(Component.text("襲撃の印", NamedTextColor.GOLD));
        meta.lore(List.of(
                Component.text("防衛戦を開始するための印", NamedTextColor.GRAY),
                Component.text("ステージ" + stageLevel + "用", NamedTextColor.GRAY)));
        item.setItemMeta(meta);
        return item;
    }

    private Optional<RaidSealItemIdentity> read(PersistentDataContainer data) {
        Byte marker = data.get(markerKey, PersistentDataType.BYTE);
        Integer version = data.get(versionKey, PersistentDataType.INTEGER);
        String sealId = data.get(sealIdKey, PersistentDataType.STRING);
        Long stage = data.get(stageLevelKey, PersistentDataType.LONG);
        if (marker == null || marker != 1 || version == null || version != ITEM_VERSION
                || sealId == null || stage == null) {
            return Optional.empty();
        }
        try {
            StageWaveSchedule.requireValidStageLevel(stage);
        } catch (IllegalArgumentException invalidStage) {
            return Optional.empty();
        }
        try {
            return Optional.of(new RaidSealItemIdentity(UUID.fromString(sealId), stage));
        } catch (IllegalArgumentException invalidId) {
            return Optional.empty();
        }
    }

    private static ItemMeta requireMeta(ItemStack item) {
        return Objects.requireNonNull(item.getItemMeta(), "raid seal metadata");
    }

    private static long requireStage(long stageLevel) {
        return StageWaveSchedule.requireValidStageLevel(stageLevel);
    }

    private static boolean isSupportedMaterial(Material material) {
        return RaidSealMaterialPolicy.supports(material.name());
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/paper/RaidSealTagger.kt
package io.github.takenoha.towerdefense.paper

import io.github.takenoha.towerdefense.domain.StageWaveSchedule
import java.util.Objects
import java.util.Optional
import java.util.OptionalLong
import java.util.UUID
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.ItemMeta
import org.bukkit.persistence.PersistentDataContainer
import org.bukkit.persistence.PersistentDataType
import org.bukkit.plugin.Plugin

/** Creates and validates the non-stackable, database-backed raid-start token. */
class RaidSealTagger(plugin: Plugin) {
    private val markerKey: NamespacedKey
    private val versionKey: NamespacedKey
    private val sealIdKey: NamespacedKey
    private val stageLevelKey: NamespacedKey

    init {
        Objects.requireNonNull(plugin, "plugin")
        markerKey = NamespacedKey(plugin, "raid_seal")
        versionKey = NamespacedKey(plugin, "raid_seal_version")
        sealIdKey = NamespacedKey(plugin, "raid_seal_id")
        stageLevelKey = NamespacedKey(plugin, "raid_seal_stage")
    }

    /** Template used by the stage-1 recipe; a craft event replaces its empty UUID marker. */
    fun recipeTemplate(): ItemStack = recipeTemplate(FOUNDATION_STAGE)

    /** Template used by a stage-specific recipe before craft-time UUID registration. */
    fun recipeTemplate(stageLevel: Long): ItemStack = createTemplate(stageLevel)

    fun create(sealId: UUID, stageLevel: Long): ItemStack {
        Objects.requireNonNull(sealId, "sealId")
        requireStage(stageLevel)
        val item = createTemplate(stageLevel)
        val meta = requireMeta(item)
        val data = meta.persistentDataContainer
        data.set(sealIdKey, PersistentDataType.STRING, sealId.toString())
        data.set(stageLevelKey, PersistentDataType.LONG, stageLevel)
        meta.displayName(Component.text("襲撃の印", NamedTextColor.GOLD))
        meta.lore(
            listOf(
                Component.text("ステージ${stageLevel}の防衛戦を開始します", NamedTextColor.GRAY),
                Component.text("右クリックでチームのコアへ使用", NamedTextColor.GRAY),
                Component.text("真正性はサーバーの永続データで検証されます", NamedTextColor.DARK_GRAY),
            )
        )
        item.itemMeta = meta
        item.amount = 1
        return item
    }

    fun read(item: ItemStack?): Optional<RaidSealItemIdentity> {
        if (item == null || !isSupportedMaterial(item.type) || item.amount != 1) {
            return Optional.empty()
        }
        val meta = item.itemMeta
        return if (meta == null) Optional.empty() else read(meta.persistentDataContainer)
    }

    /** Returns whether a valid seal still uses the pre-UX ENDER_EYE material. */
    fun isLegacyMaterial(item: ItemStack?): Boolean =
        item != null && item.type == LEGACY_ITEM_MATERIAL && read(item).isPresent

    fun isRecipeTemplate(item: ItemStack?): Boolean = templateStage(item).isPresent

    /** Reads the stage encoded in a registered recipe result. */
    fun templateStage(item: ItemStack?): OptionalLong {
        if (item == null || !isSupportedMaterial(item.type) || item.amount != 1) {
            return OptionalLong.empty()
        }
        val meta = item.itemMeta ?: return OptionalLong.empty()
        val data = meta.persistentDataContainer
        val version = data.get(versionKey, PersistentDataType.INTEGER)
        val marker = data.get(markerKey, PersistentDataType.BYTE)
        if (marker == null || marker != 1.toByte() || version == null ||
            version != ITEM_VERSION || data.get(sealIdKey, PersistentDataType.STRING) != null
        ) {
            return OptionalLong.empty()
        }
        val stage = data.get(stageLevelKey, PersistentDataType.LONG)
            ?: return OptionalLong.empty()
        return try {
            StageWaveSchedule.requireValidStageLevel(stage)
            OptionalLong.of(stage)
        } catch (_: IllegalArgumentException) {
            OptionalLong.empty()
        }
    }

    fun hasSealId(item: ItemStack?, sealId: UUID): Boolean {
        Objects.requireNonNull(sealId, "sealId")
        return read(item).map { identity -> identity.sealId == sealId }.orElse(false)
    }

    private fun createTemplate(stageLevel: Long): ItemStack {
        requireStage(stageLevel)
        val item = ItemStack(ITEM_MATERIAL, 1)
        val meta = requireMeta(item)
        val data = meta.persistentDataContainer
        data.set(markerKey, PersistentDataType.BYTE, 1.toByte())
        data.set(versionKey, PersistentDataType.INTEGER, ITEM_VERSION)
        data.set(stageLevelKey, PersistentDataType.LONG, stageLevel)
        meta.displayName(Component.text("襲撃の印", NamedTextColor.GOLD))
        meta.lore(
            listOf(
                Component.text("防衛戦を開始するための印", NamedTextColor.GRAY),
                Component.text("ステージ${stageLevel}用", NamedTextColor.GRAY),
            )
        )
        item.itemMeta = meta
        return item
    }

    private fun read(data: PersistentDataContainer): Optional<RaidSealItemIdentity> {
        val marker = data.get(markerKey, PersistentDataType.BYTE)
        val version = data.get(versionKey, PersistentDataType.INTEGER)
        val sealId = data.get(sealIdKey, PersistentDataType.STRING)
        val stage = data.get(stageLevelKey, PersistentDataType.LONG)
        if (marker == null || marker != 1.toByte() || version == null ||
            version != ITEM_VERSION || sealId == null || stage == null
        ) {
            return Optional.empty()
        }
        try {
            StageWaveSchedule.requireValidStageLevel(stage)
        } catch (_: IllegalArgumentException) {
            return Optional.empty()
        }
        return try {
            Optional.of(RaidSealItemIdentity(UUID.fromString(sealId), stage))
        } catch (_: IllegalArgumentException) {
            Optional.empty()
        }
    }

    private fun requireMeta(item: ItemStack): ItemMeta =
        Objects.requireNonNull(item.itemMeta, "raid seal metadata")

    private fun requireStage(stageLevel: Long): Long =
        StageWaveSchedule.requireValidStageLevel(stageLevel)

    private fun isSupportedMaterial(material: Material): Boolean =
        RaidSealMaterialPolicy.supports(material.name)

    companion object {
        const val ITEM_VERSION: Int = 1
        const val FOUNDATION_STAGE: Long = 1L

        @JvmField
        val ITEM_MATERIAL: Material = Material.valueOf(RaidSealMaterialPolicy.CURRENT_MATERIAL)

        @JvmField
        val LEGACY_ITEM_MATERIAL: Material = Material.valueOf(RaidSealMaterialPolicy.LEGACY_MATERIAL)
    }
}
